client/app/scripts/utils/update-buffer-utils.js

```diff
@@ -0,0 +1,129 @@
+import AppStore from '../stores/app-store';
+
+const log = debug('scope:update-buffer-utils');
+const makeList = Immutable.List;
+const feedInterval = 1000;
+const bufferLength = 100;
+
+let deltaBuffer = makeList();
+let detailsBuffer = makeList();
+let updateTimer = null;
+
+function isPaused() {
+  return AppStore.isUpdatePaused();
+}
+
+export function resetDetailsUpdateBuffer() {
+  detailsBuffer = detailsBuffer.clear();
+}
+
+export function resetUpdateBuffer() {
+  clearTimeout(updateTimer);
+  deltaBuffer = deltaBuffer.clear();
+  detailsBuffer = detailsBuffer.clear();
+}
+
+function maybeUpdate() {
+  if (isPaused()) {
+    clearTimeout(updateTimer);
+    resetUpdateBuffer();
+  } else {
+    if (deltaBuffer.size > 0) {
+      const delta = deltaBuffer.first();
+      deltaBuffer = deltaBuffer.shift();
+      receiveNodesDelta(delta);
+    }
+    if (detailsBuffer.size > 0) {
+      const details = detailsBuffer.first();
+      detailsBuffer = detailsBuffer.shift();
+      receiveNodeDetails(details);
+    }
+    if (detailsBuffer.size > 0 || deltaBuffer.size > 0) {
+      updateTimer = setTimeout(maybeUpdate, feedInterval);
+    }
+  }
+}
+
+export function bufferDeltaUpdate(delta) {
+  if (delta.add === null && delta.update === null && delta.remove === null) {
+    log('Discarding empty nodes delta');
+    return;
+  }
+
+  if (deltaBuffer.size >= bufferLength) {
+    // consolidate first buffer entry with second
+    const first = deltaBuffer.first();
+    deltaBuffer = deltaBuffer.shift();
+    const second = deltaBuffer.first();
+    let toAdd = _.union(first.add, second.add);
+    let toUpdate = _.union(first.update, second.update);
+    let toRemove = _.union(first.remove, second.remove);
+    log('Consolidating delta buffer', 'add', _.size(toAdd), 'update', _.size(toUpdate), 'remove', _.size(toRemove));
+
+    // check if an added node in first was updated in second -> add second update
+    toAdd = _.map(toAdd, node => {
+      const updateNode = _.find(second.update, {'id': node.id});
+      if (updateNode) {
+        toUpdate = _.reject(toUpdate, {'id': node.id});
+        return updateNode;
+      }
+      return node;
+    });
+
+    // check if an updated node in first was updated in second -> updated second update
+    // no action needed, successive updates are fine
+
+    // check if an added node in first was removed in second -> dont add, dont remove
+    _.each(first.add, node => {
+      const removedNode = _.find(second.remove, {'id': node.id});
+      if (removedNode) {
+        toAdd = _.reject(toAdd, {'id': node.id});
+        toRemove = _.reject(toRemove, {'id': node.id});
+      }
+    });
+
+    // check if an updated node in first was removed in second -> remove
+    _.each(first.update, node => {
+      const removedNode = _.find(second.remove, {'id': node.id});
+      if (removedNode) {
+        toUpdate = _.reject(toUpdate, {'id': node.id});
+      }
+    });
```

Comment:
Maybe pull these chunks out into a "consolidateDeltas" fn, I know I'm lazy but I kind of skipped over why we were doing all this consolidation stuff on every buffered delta w/out realising this only happens if we fill up the buffer.